pandas/core/indexes/datetimelike.py

```diff
@@ -602,6 +602,15 @@ def _shallow_copy(self, values=None, name: Label = lib.no_default):
         result._cache = cache
         return result
 
+    def factorize(self, sort=False, na_sentinel=-1):
+        if self.freq is not None and sort is False:
+            # we are unique, so can short-circuit, also can preserve freq
+            codes = np.arange(len(self))
+            return codes, self[:]
+            # TODO: In the sort=True case we could check for montonic_decreasing
+            #  and operate on self[::-1]
+        return super().factorize(sort=sort, na_sentinel=na_sentinel)
```

Comment:
This seems to be duplicating the array version, can't that be reused? (directly, or by having `pd.factorize` call the array version)

Reply:
Can you check this comment?

Reply:
Its close; the main difference is that the Index version returns an Index for uniques

Reply:
But that is handled in `factorize` (depending on the input, it will wrap the uniques in an Index or not)?

Reply:
im not clear on what youre suggesting?  that we dont need to override this here at all?

Reply:
yes

Reply:
tried that, it breaks the tests for checking that the freq is preserved

Reply:
Well, you might need to update `factorize` to preserve `freq` (it's using `_shallow_copy` right now)

(or we could decide that `factorize` is not an operation that should preserve the freq)

Reply:
> Well, you might need to update factorize to preserve freq (it's using _shallow_copy right now)

If we're getting into Index-subclass-specific logic, I think that belongs on the Index subclass.

Reply:
> (or we could decide that factorize is not an operation that should preserve the freq)

I'd be OK with this

Reply:
> If we're getting into Index-subclass-specific logic, I think that belongs on the Index subclass.

Not necessarily, I would say, as it already has index-specific handling as well. 
But there is no "shallow_copy"-like method that preserves attributes like `freq` ?

Reply:
>  But there is no "shallow_copy"-like method that preserves attributes like freq ?

There used to be, but it was being used in places where freq shouldnt be retained, so was changed as it was a footgun.